src/ribasim_nl/ribasim_nl/network_validator.py

```diff
@@ -0,0 +1,105 @@
+from dataclasses import dataclass
+
+from ribasim import Model
+
+
+def within_distance(row, gdf, tolerance=1.0) -> bool:
+    distance = gdf[gdf.index != row.name].distance(row.geometry)
+    return (distance < tolerance).any()
+
+
+def check_node_connectivity(row, node_df, tolerance=1.0) -> bool:
+    invalid = True
+
+    # check if from_node_id is valid
+    if row.from_node_id in node_df.index:
+        distance = row.geometry.boundary.geoms[0].distance(node_df.at[row.from_node_id, "geometry"])
+        invalid = distance > tolerance
+
+    # if valid, check if to_node_id is valid
+    if (not invalid) and (row.to_node_id in node_df.index):
+        distance = row.geometry.boundary.geoms[1].distance(node_df.at[row.to_node_id, "geometry"])
+        invalid = distance > tolerance
+
+    return invalid
+
+
+def check_internal_basin(row, edge_df) -> bool:
+    if row.node_type == "Basin":
+        return row.node_id not in edge_df.from_node_id.to_numpy()
+    else:
+        return False
+
+
+@dataclass
+class NetworkValidator:
+    model: Model
+    tolerance: float = 1
```

Comment:
Needs a short docstring. Also good to note that a tolerance of 1 is huge if lat long coordinates are used.

Reply:
Put a docstring, including the comment tolerance should be expressed in units of the model crs.